pkg/genericapiserver/enablement_config.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package genericapiserver
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// APIResourceEnablement is the interface to determine which versions and resources are enabled
+type APIResourceEnablement interface {
+	ResourceEnabled(resource unversioned.GroupVersionResource) bool
+	VersionEnabled(version unversioned.GroupVersion) bool
+}
+
+// Specifies the overrides for various API group versions.
+// This can be used to enable/disable entire group versions or specific resources.
+type APIGroupVersionEnablementConfig struct {
+	// Whether to enable or disable this entire group version.  This dominates any enablement check.
+	Enable bool
+
+	// DisabledResources lists the resources that are specifically disabled for a group/version
+	// DisabledResources trumps EnabledResources
+	DisabledResources sets.String
+
+	// EnabledResources lists the resources that should be enabled by default.  This is a little
+	// unusual, but we need it for compatibility with old code for now.  An empty set means
+	// enable all, a non-empty set means that all other resources are disabled.
+	EnabledResources sets.String
+}
+
+var _ APIResourceEnablement = &APIResourceEnablementConfig{}
+
+// AnyVersion is used during resource enablement checks to see if any version of the resource is enabled
+const AnyVersion = "."
```

Comment:
OK, I think I'd still really prefer a separate function rather than the
magic constant--I think implementers are more likely to get it right.

On Tue, Mar 22, 2016 at 1:32 PM, David Eads <notifications@github.com>
wrote:

> In pkg/genericapiserver/enablement_config.go
> <https://github.com/kubernetes/kubernetes/pull/23067#discussion_r57063586>
> :
>
> > +	Enable bool
> > +
> > +	// DisabledResources lists the resources that are specifically disabled for a group/version
> > +	// DisabledResources trumps EnabledResources
> > +	DisabledResources sets.String
> > +
> > +	// EnabledResources lists the resources that should be enabled by default.  This is a little
> > +	// unusual, but we need it for compatibility with old code for now.  An empty set means
> > +	// enable all, a non-empty set means that all other resources are disabled.
> > +	EnabledResources sets.String
> > +}
> > +
> > +var _ APIResourceEnablement = &APIResourceEnablementConfig{}
> > +
> > +// AnyVersion is used during resource enablement checks to see if any version of the resource is enabled
> > +const AnyVersion = "."
>
> Actually I see that you don't actually call with this anywhere. Can you
> just omit the functionality until we have a use case?
>
> The next pull that fixes StorageDestinations uses this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23067/files/cdef546dee0c2d222b6ca0cfee281079b5be2ce6#r57063586>
>